src/index.js

```diff
@@ -850,42 +879,54 @@ export default class Carousel extends React.Component {
     };
   }
 
+  getSlideDirection(start, end, isWrapping) {
+    let direction = 0;
+    if (start === end) return direction;
+
+    if (isWrapping) {
+      direction = start < end ? -1 : 1;
+    } else {
+      direction = start < end ? 1 : -1;
+    }
+
+    return direction;
+  }
+
   getSlideTargetPosition(index, positionValue) {
-    const slidesToShow = this.state.frameWidth / this.state.slideWidth;
-    const targetPosition =
+    let targetPosition =
       (this.state.slideWidth + this.props.cellSpacing) * index;
-    const end =
-      (this.state.slideWidth + this.props.cellSpacing) * slidesToShow * -1;
-
-    if (
-      this.props.wrapAround &&
-      (this.state.isWrappingAround || this.state.dragging)
-    ) {
-      const slidesBefore = Math.ceil(positionValue / this.state.slideWidth);
-      if (this.state.slideCount - slidesBefore <= index) {
-        return (
-          (this.state.slideWidth + this.props.cellSpacing) *
-          (this.state.slideCount - index) *
-          -1
-        );
-      }
+    const startSlide = Math.min(
+      Math.abs(Math.floor(positionValue / this.state.slideWidth)),
+      this.state.slideCount - 1
+    );
 
-      let slidesAfter = Math.ceil(
-        (Math.abs(positionValue) - Math.abs(end)) / this.state.slideWidth
+    if (this.props.wrapAround && index !== startSlide) {
+      const direction = this.getSlideDirection(
+        startSlide,
+        this.state.currentSlide,
+        this.state.isWrappingAround
       );
+      let slidesBefore = Math.floor((this.state.slideCount - 1) / 2);
+      let slidesAfter = this.state.slideCount - slidesBefore - 1;
 
-      if (this.state.slideWidth !== 1) {
-        slidesAfter = Math.ceil(
-          (Math.abs(positionValue) - this.state.slideWidth) /
-            this.state.slideWidth
-        );
+      if (direction < 0) {
+        const temp = slidesBefore;
+        slidesBefore = slidesAfter;
+        slidesAfter = temp;
```

Comment:
This totally fixes #194 and #340. Good job getting the bigger half of the slides to show in the direction that it's sliding.